sources/host-ctr/pidnamespaces/pid_namespaces.go

```diff
@@ -0,0 +1,169 @@
+package pidnamespaces
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/gofrs/flock"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	lockfile     string = "/run/lock/host-ctr-pid-ns.lock"
+	manifestPath string = "/etc/host-containers/info/pid-namespaces.json"
+)
+
+// HostCtrPidNs represents the list of PID namespaces
+type HostCtrPidNs struct {
+	PidNamespaces []int64 `json:"pid-namespaces,omitempty"`
+}
+
+// GetTaskPidNs returns the PID NS associated with the container task PID or error on failure
+func GetTaskPidNs(pid uint32) (int64, error) {
+	pidNsFile, err := os.Open(fmt.Sprintf("/proc/%d/ns/pid", pid))
+	if err != nil {
+		return -1, fmt.Errorf("failed to open %d PID namespace file", pid)
+	}
+	pidNsFd := int(pidNsFile.Fd())
+
+	// Get file descriptor that refers to a parent namespace, ioctl() request '_IO(NSIO, 0x2)'
+	_, err = unix.IoctlGetInt(pidNsFd, unix.NS_GET_PARENT)
+	if err == unix.EPERM {
+		// If we can't access the parent PID namespace, this means the current container task is sharing the root PID
+		// namespace. We don't want to track the root PID namespace in the manifest, so we return -1.
+		return -1, nil
+	} else if err != nil {
+		return -1, errors.Wrap(err, fmt.Sprintf("ioctl failed to 'NS_GET_PARENTNS' for '%s'", pidNsFile.Name()))
+	}
+
+	// The PID namespace is represented by the inode number of the task's PID namespace file
+	var sb unix.Stat_t
+	if err := unix.Fstat(pidNsFd, &sb); err != nil {
+		return -1, errors.Wrap(err, "failed to stat task PID namespace fd")
+	}
+
+	return int64(sb.Ino), nil
+}
+
+// AddPidNsToManifest takes a given PID's namespace and adds it to the internal manifest
+func AddPidNsToManifest(pidNs int64) (err error) {
+	fileLock := flock.New(lockfile)
+	lockCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	ok, err := fileLock.TryLockContext(lockCtx, 500*time.Millisecond)
+	if err != nil && !ok {
+		return errors.Wrap(err, "acquiring lock for writing to PID NS manifest failed with error")
+	}
+	defer func() {
+		if err = fileLock.Unlock(); err != nil {
+			err = errors.Wrap(err, "failed to release lock for writing to PID NS manifest")
+		}
+	}()
+
+	// Add task PID NS to manifest
+	manifest, err := os.OpenFile(manifestPath, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return errors.Wrap(err, "failed to open PID NS manifest file")
+	}
+	defer manifest.Close()
+
+	manifestStat, err := manifest.Stat()
+	if err != nil {
+		return errors.Wrap(err, "failed to stat PID NS manifest file")
+	}
+	raw := make([]byte, manifestStat.Size())
+	if _, err = manifest.Read(raw); err != nil {
+		return errors.Wrap(err, "failed to read PID NS manifest file")
+	}
```

Comment:
This has a TOCTOU issue since the file might grow between the `stat` and `read` calls.

Is there not a helper that reads all bytes from a file into a `[]byte` without the stat call?

Reply:
@jpmcb called out the same thing in https://github.com/bottlerocket-os/bottlerocket/pull/2839#discussion_r1120911348. My reasoning was that I already had an open fd I can use without having to call `ReadFile` which opens the file again for reading. There is no utility in `os.File` that lets you read all the bytes. You can only call `os.ReadFile` for that.

In terms of TOCTOU, it is still a concern if there can only be one `host-ctr` process opening/touching this file?

Reply:
After the refactor, I ended up using `os.ReadFile` but it doesn't avoid the `stat` and `read`.